### PR TITLE
feat(Core/PacketIO): Implement STATUS_LOGGEDIN_OR_RECENTLY_LOGGOUT

### PR DESCRIPTION
**Pull request by Kitzunu**
* Cherry-pick commit (https://github.com/TrinityCore/TrinityCore/commit/dc68c2b269fdd401e2f0f6e1e372292b74c66bb8)
* Cherry-pick commit (https://github.com/TrinityCore/TrinityCore/commit/9fc5c4bc41b2738836ce99eb42ddbb88fb398884)
* cherry-pick commit (https://github.com/TrinityCore/TrinityCore/commit/f1e6601032ba4077e1d1c6070c4cb20812289f7a)
* cherry-pick commit (https://github.com/TrinityCore/TrinityCore/commit/aee4f0350ea5610b663cb07fb097f2fad256809e)

Funny that someone have cherry-picked 90% of these commits already, but forgot about the important things x)

## Changes Proposed:
- Implement m_playerRecentlyLogout & STATUS_LOGGEDIN_OR_RECENTLY_LOGGOUT

- make use of requeuePackets

* For client packets that can be send in loggined state or just after logout but assocualted with recently logout character add new login status STATUS_LOGGEDIN_OR_RECENTLY_LOGGOUT

* Requeue packets handled with STATUS_LOGGEDIN received before player is fully loaded in one step after packet processing loop - reduces used locks from 3 per packet to 1 per packet

* Allow CMSG_LOGOUT_CANCEL to be recieved during serverside logout process (ignore it however, it's too late to abort)

* Fix STATUS_LOGGEDIN_OR_RECENTLY_LOGGOUT opcodes because of m_playerRecentlyLogout being set false too early in some cases

## Issues Addressed:

- Closes 

## Tests Performed:

- Test cancel trade
- Test cancel logout

## How to Test the Changes:

1. Tested opcodes in-game with LOG_ERROR to know that we recieve the opcodes

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Server/WorldSession.cpp [hunk 1]
```diff
@@ -326,21 +327,18 @@ bool WorldSession::Update(uint32 diff, PacketFilter& updater)
...
+                        requeuePackets.push_back(packet);
                         deletePacket = false;
                         QueuePacket(packet);
-                    }*/
+
+                        LOG_DEBUG("network", "Re-enqueueing packet with opcode %s with with status STATUS_LOGGEDIN. "
+                                    "Player {} is currently not in world yet.", GetOpcodeNameForLogging(static_cast<OpcodeClient>(packet->GetOpcode())), GetPlayerInfo());
+                    }
                 }
                 else if (_player->IsInWorld() && AntiDOS.EvaluateOpcode(*packet, currentTime))
```
Comment (Roxtedy) [inline src/server/game/Server/WorldSession.cpp:343]: Should we continue processing packages? How is it here TrinityCore/TrinityCore@aee4f0350ea5610b663cb07fb097f2fad256809e

Comment (Kitzunu) [reply src/server/game/Server/WorldSession.cpp:343]: Ah yes, this is true

**Review by Nyeriah — APPROVED** [review]

**Comment by hahalmtdhl** [comment]
Have you fixed this problem?

**Comment by Kitzunu** [comment]
Yes thats why the issue is closed

**Comment by hahalmtdhl** [comment]
okay, thank you.

**Pull request merged**